worktree_commit.go

```diff
@@ -45,29 +45,37 @@ func (w *Worktree) Commit(msg string, opts *CommitOptions) (plumbing.Hash, error
 		if err != nil {
 			return plumbing.ZeroHash, err
 		}
-
-		t, err := w.r.getTreeFromCommitHash(head.Hash())
+		headCommit, err := w.r.CommitObject(head.Hash())
 		if err != nil {
 			return plumbing.ZeroHash, err
 		}
 
-		treeHash = t.Hash
-		opts.Parents = []plumbing.Hash{head.Hash()}
-	} else {
-		idx, err := w.r.Storer.Index()
-		if err != nil {
-			return plumbing.ZeroHash, err
+		opts.Parents = nil
+		if len(headCommit.ParentHashes) != 0 {
+			opts.Parents = []plumbing.Hash{headCommit.ParentHashes[0]}
 		}
 
-		h := &buildTreeHelper{
-			fs: w.Filesystem,
-			s:  w.r.Storer,
-		}
+		// After everything goes well, should delete previous head commit
+		defer func(err *error) {
+			if *err == nil {
+				*err = w.r.DeleteObject(head.Hash())
+			}
+		}(&err)
```

Comment:
There is no need to delete the object as it will simply become unreachable.

```suggestion
```

Reply:
I'm not familiar with this codebase. I was thinking that there could be memory leak or waste of space if I don't delete it. Is there any way to avoid it or is it ignorable?

Reply:
@pjbgf I'm still curious!